hello-jniCallback/app/src/main/jni/hello-jnicallback.c

```diff
@@ -0,0 +1,314 @@
+
+// processing callback to handler class
+typedef struct tick_context {
+    JavaVM  *javaVM;
+    jclass   jniHelperClz;
+    jobject  jniHelperObj;
+    jclass   mainActivityClz;
+    jobject  mainActivityObj;
+    pthread_mutex_t  lock;
+    int      done;
+} TickContext;
+TickContext g_ctx;
+
+/* This is a trivial JNI example where we use a native method
+ * to return a new VM String. See the corresponding Java source
+ * file located at:
+ *
+ *   apps/samples/hello-jni/project/src/com/example/hellojni/HelloJni.java
+ */
+jstring
+Java_com_example_hello_1jnicallback_MainActivity_stringFromJNI( JNIEnv* env, jobject thiz )
+{
+#if defined(__arm__)
+  #if defined(__ARM_ARCH_7A__)
+    #if defined(__ARM_NEON__)
+      #if defined(__ARM_PCS_VFP)
+        #define ABI "armeabi-v7a/NEON (hard-float)"
+      #else
+        #define ABI "armeabi-v7a/NEON"
+      #endif
+    #else
+      #if defined(__ARM_PCS_VFP)
+        #define ABI "armeabi-v7a (hard-float)"
+      #else
+        #define ABI "armeabi-v7a"
+      #endif
+    #endif
+  #else
+   #define ABI "armeabi"
+  #endif
+#elif defined(__i386__)
+   #define ABI "x86"
+#elif defined(__x86_64__)
+   #define ABI "x86_64"
+#elif defined(__mips64)  /* mips64el-* toolchain defines __mips__ too */
+   #define ABI "mips64"
+#elif defined(__mips__)
+   #define ABI "mips"
+#elif defined(__aarch64__)
+   #define ABI "arm64-v8a"
+#else
+   #define ABI "unknown"
+#endif
+    return (*env)->NewStringUTF(env, "Hello from JNI !  Compiled with ABI " ABI ".");
+}
+
+/*
+ *  A helper function to show how to call
+ *     java static functions JniHelper::getBuildVersion()
+ *     java non-static function JniHelper::getRuntimeMemorySize()
+ *  The trivial implementation for these functions are inside file
+ *     JniHelper.java
+ */
+void queryRuntimeInfo(JNIEnv *env, jobject instance) {
+    // Find out which OS we are running on. It does not matter for this app
+    // just to demo how to call static functions.
+    // Our java JniHelper class id and instance are initialized when this
+    // shared lib got loaded, we just directly use them
+    //    static function does not need instance, so we just need to feed
+    //    class and method id to JNI
+    jmethodID versionFunc = (*env)->GetStaticMethodID(
+            env, g_ctx.jniHelperClz,
+            "getBuildVersion", "()Ljava/lang/String;");
+    if (!versionFunc) {
+        LOGE("Failed to retrieve getBuildVersion() methodID @ line %d",
+             __LINE__);
+        return;
+    }
+    jstring buildVersion = (*env)->CallStaticObjectMethod(env,
+                                       g_ctx.jniHelperClz, versionFunc);
+    const char *version = (*env)->GetStringUTFChars(env, buildVersion, NULL);
+    if (!version) {
+        LOGE("Unable to get version string @ line %d", __LINE__);
+        return;
+    }
+    LOGI("Android Version - %s", version);
+    (*env)->ReleaseStringUTFChars(env, buildVersion, version);
+
```

Comment:
As the thread is attached via AttachCurrentThread(), the VM won't release local references as described below. So we need to release it?

https://developer.android.com/training/articles/perf-jni.html
>One unusual case deserves separate mention. If you attach a native thread with AttachCurrentThread, the code you are running will never automatically free local references until the thread detaches. Any local references you create will have to be deleted manually. In general, any native code that creates local references in a loop probably needs to do some manual deletion.

I'm fine with the sample following C style now. In future we want to migrate all C samples to C++ ?